eclair-core/src/main/scala/fr/acinq/eclair/io/MessageRelay.scala

```diff
@@ -88,25 +102,44 @@ object MessageRelay {
         replyTo_opt.foreach(_ ! UnknownOutgoingChannel(messageId, outgoingChannelId))
         Behaviors.stopped
       case (context, WrappedOptionalNodeId(Some(nextNodeId))) =>
-        withNextNodeId(context, messageId, switchboard, prevNodeId, nextNodeId, msg, policy, replyTo_opt)
+        withNextNodeId(context, nodeParams, messageId, switchboard, register, router, prevNodeId, nextNodeId, msg, policy, replyTo_opt)
     }
 
   def withNextNodeId(context: ActorContext[Command],
+                     nodeParams: NodeParams,
                      messageId: ByteVector32,
                      switchboard: ActorRef,
+                     register: ActorRef,
+                     router: typed.ActorRef[Router.GetNodeId],
                      prevNodeId: PublicKey,
                      nextNodeId: PublicKey,
                      msg: OnionMessage,
                      policy: RelayPolicy,
-                     replyTo_opt: Option[typed.ActorRef[Status]]): Behavior[Command] =
-        policy match {
-          case RelayChannelsOnly =>
-            switchboard ! GetPeerInfo(context.messageAdapter(WrappedPeerInfo), prevNodeId)
-            waitForPreviousPeer(messageId, switchboard, nextNodeId, msg, replyTo_opt)
-          case RelayAll =>
-            switchboard ! Peer.Connect(nextNodeId, None, context.messageAdapter(WrappedConnectionResult).toClassic, isPersistent = false)
-            waitForConnection(messageId, msg, replyTo_opt)
-        }
+                     replyTo_opt: Option[typed.ActorRef[Status]]): Behavior[Command] = {
+    if (nextNodeId == nodeParams.nodeId) {
+      OnionMessages.process(nodeParams.privateKey, msg) match {
+        case OnionMessages.DropMessage(reason) =>
+          replyTo_opt.foreach(_ ! DroppedMessage(messageId, reason))
+          Behaviors.stopped
+        case OnionMessages.SendMessage(nextNode, nextMessage) =>
+          context.self ! RelayMessage(messageId, prevNodeId, nextNode, nextMessage, policy, replyTo_opt)
+          apply(nodeParams: NodeParams, switchboard, register, router)
```

Comment:
We need to more clearly name the states of the state machine here, this can be improved. I believe it's also time to factor the parameters like we do in most other actors, by using a `private class`.

Reply:
Done